captcha/tests/test_fields.py

```diff
@@ -0,0 +1,166 @@
+        self.assertIn('data-widget-uuid="%s"' % test_hex, html)
+        self.assertIn('data-sitekey="pubkey"', html)
+        self.assertIn("var onSubmit_%s = function(token) {" % test_hex, html)
+
+    @patch("captcha.widgets.uuid.UUID.hex", new_callable=PropertyMock)
+    def test_v2_checkbox_attribute_changes_html(self, mocked_uuid):
+        test_hex = "e83ccae286ad4784bd47f7ddc40cfd6f"
+        mocked_uuid.return_value = test_hex
+        class CheckboxAttrForm(forms.Form):
+            captcha = fields.ReCaptchaField(
+                widget=widgets.ReCaptchaV2Checkbox(
+                    attrs={
+                        "data-theme": "dark",
+                        "language": "af",
+                        "data-callback": "customCallback",
+                        "data-size": "compact"
+                    }
+                )
+            )
+        form = CheckboxAttrForm()
+        html = form.as_p()
+        self.assertIn('<script src="https://www.google.com/recaptcha/api.js?hl=af"></script>', html)
+        self.assertIn('data-theme="dark"', html)
+        self.assertNotIn('data-callback="onSubmit_%s"' % test_hex, html)
+        self.assertIn('data-callback="customCallback"', html)
+        self.assertIn('data-size="compact"', html)
+        self.assertIn('class="g-recaptcha"', html)
+        self.assertIn('required="True"', html)
+        self.assertIn('data-widget-uuid="%s"' % test_hex, html)
+        self.assertIn('data-sitekey="pubkey"', html)
+        self.assertIn("var onSubmit_%s = function(token) {" % test_hex, html)
+
+    @patch("captcha.widgets.uuid.UUID.hex", new_callable=PropertyMock)
+    def test_default_v2_invisible_html(self, mocked_uuid):
+        test_hex = "72f853eb8b7e4022b808be0f5c3bc297"
+        mocked_uuid.return_value = test_hex
+        class InvisForm(forms.Form):
+            captcha = fields.ReCaptchaField(
+                widget=widgets.ReCaptchaV2Invisible()
+            )
+        form = InvisForm()
+        html = form.as_p()
+        self.assertIn('<script src="https://www.google.com/recaptcha/api.js?hl=en"></script>', html)
+        self.assertIn('data-size="invisible"', html)
+        self.assertIn('data-callback="onSubmit_%s"' % test_hex, html)
+        self.assertIn('class="g-recaptcha"', html)
+        self.assertIn('required="True"', html)
+        self.assertIn('data-widget-uuid="%s"' % test_hex, html)
+        self.assertIn('data-sitekey="pubkey"', html)
+        self.assertIn("var onSubmit_%s = function(token) {" % test_hex, html)
+        self.assertIn("var verifyCaptcha_%s = function(e) {" % test_hex, html)
+        self.assertIn('.g-recaptcha[data-widget-uuid="%s"]' % test_hex, html)
+
+    @patch("captcha.widgets.uuid.UUID.hex", new_callable=PropertyMock)
+    def test_v2_invisible_attribute_changes_html(self, mocked_uuid):
+        test_hex = "8b220c54ddb849b8bb59bda5da57baea"
+        mocked_uuid.return_value = test_hex
+        class InvisAttrForm(forms.Form):
```

Comment:
Whitespace before and after class definition, please.

Reply:
As per external discussion, JavaScript and HTML improvements can be done at a later date.